mesonbuild/compilers/mixins/clike.py

```diff
@@ -241,10 +286,10 @@ def get_library_dirs(self, env, elf_class = None):
                     # Skip the file if we can't read it
                     pass
 
-        return tuple(retval)
+        return retval
```

Comment:
If we're using lru_cache, then we should either return a tuple (or some other immutable sequence) or make sure that everything that consumes it copies before mutating the list, otherwise the cached value will change.

Reply:
Actually, tuples are mutable too. Really we need to copy the return value to ensure that if a caller mutates it, then they won't cause problems. Or document that callers need to copy before mutation. I'll double check that before the next version

Reply:
The tuple type is a container which is not mutable, but it can contain other mutable types. Tuples of strings are completely immutable. I think it's easier to assert that all functions that return a string sequence are tuples using mypy than to assert that they are returning _a copy of a list_. So I think it would be safer to always return a tuple from all our compiler APIs.

Reply:
:/ It's just really gross to treat tuples as immutable lists, because that's not what they're for.

Reply:
I've wrapped the cached function as a private/public pair. If you're deadset on using a tuple I guess we can do that, but I really dislike using `tuple[str, ...]`, because that's a code smell.

Reply:
We really need some kind of immutable sequence type to return from all our compiler methods. We've hit the 'oh no we accidentally returned a mutable copy and changed it for everyone' issue too many times over the years. Most recently: https://github.com/mesonbuild/meson/pull/6852, and we'll definitely hit it again.

We can do it later (i.e., not in this PR), but Typing seems like an ideal opportunity to fix this.

Reply:
As a pythonista... It isn't really possible, python is all mutable all the time. returning copies really is the best we can do. I really wish mypy had a `const[]` annotation so we could catch people mutating stuff we don't want them to without the runtime costs of using properties for that.

Reply:
There is `Final` in [typing_extensions](https://mypy.readthedocs.io/en/stable/final_attrs.html)...

Reply:
Which isn't quite what I want unfortunately, `Final` allows you to say "don't let a subclass overwrite this", so you can do things like:

```python
class Base:
   cache: Final[T.Dict[str, str]] = {}  # this needs to be shared by all subclasses

class Child(Base):
    cache: T.Dict[str, str] = {}   # mypy error!
```

Reply:
Actually, I just had a thought. We probably could get mypy to do this with a Protocol, if we had a list-like protocol that lacked any of the mutation methods....

Reply:
Yes, in my testing that seems to work... I'm going to play with that some more.

Reply:
❤